File: LeetCode/33-Search in Rotated Sorted Array.py
# 核心思路依然是二分法，只是需要注意在分的时候因为有一部分旋转位置的存在，判断处在哪一段的时候还需要先检查这一段是否完全是升序排列
# 检查方法也很简单，对比首尾元素即可，因为旋转的那一部分是从前面挪到后面的，所以如果不是完全升序，最后一个元素肯定比第一个元素小
# 然后其他的按照二分法算索引即可
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        result = -1
        
        length = len(nums)
        if (length < 1):
            return result
        
        check_index = length // 2
        check_num = nums[check_index]
        if (check_num == target):
            result = check_index
        elif (check_num >= nums[0]):
            if (target >= nums[0] and target < check_num):
                result = self.search(nums[:check_index], target)
            elif (target <= nums[-1] or target > check_num):
                sub_result = self.search(nums[check_index+1:], target)
                if (sub_result != -1):
                    result = check_index + 1 + sub_result
        else:
            if (check_num < target and target <= nums[-1]):
                sub_result = self.search(nums[check_index+1:], target)
                if (sub_result != -1):
                    result = check_index + 1 + sub_result
            elif (target < check_num or target >= nums[0]):
                result = self.search(nums[:check_index], target)
        
        return result

File: LeetCode/33. Search in Rotated Sorted Array.cpp
// 跟有序的二分思路是一样的，唯一的区别是判断的时候还要注意对应区间是否可能存在旋转的情况，可以通过当前节点和两边端点是否保持有序性来判断
// 最佳答案是先找到旋转点，然后划分旋转点之后在两边分别找，虽然代码更麻烦，但是思路更清晰，确实更好
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int start = 0;
        int end = nums.size();
        int mid;
        int ans = -1;
        while(start < end) {
            mid = (start + end) / 2;
            if (nums[mid] == target) {
                return mid;
            }
            else if (nums[mid] < target) {
                if (nums[mid] > nums[nums.size() - 1] || target <= nums[nums.size() - 1]) {
                    start = mid + 1;
                }
                else {
                    end = mid;
                }
            }
            else {
                if (target >= nums[0] || nums[mid] < nums[0]) {
                    end = mid;
                }
                else {
                    start = mid + 1;
                }
            }
        }
        
        return ans;
    }
};